In [1]:
import re

In [2]:
class Piece:
    def __init__(self, s):
        lines = s.split("\n")
        self.id = re.match(r"(\d+):", lines[0]).group(1)
        self.grid = [c == '#' for c in "".join(lines[1:])]
    def numpoints(self):
        return sum(self.grid)

class Region:
    def __init__(self, s):
        integers = [int(s) for s in re.findall(r'\d+', s)]
        self.desc = s
        self.x = integers[0]
        self.y = integers[1]
        self.pieces = integers[2:]

sections = open("./data/input.txt", "r").read().strip().split("\n\n")
pieces = [Piece(s) for s in sections[:-1]]
regions = [Region(line) for line in sections[-1].split("\n")]

In [11]:
# Check the number of 3x3 slots of the region vs number of pieces

print("Found {} pieces".format(len(pieces)))
print("Found {} regions".format(len(regions)))

regions_too_small = [r for r in regions if (r.x // 3) * (r.y // 3) < sum(r.pieces)]
print("Of which {} regions have too many pieces to fit in 3x3 blocks".format(len(regions_too_small)))

gaps = [(sum(r.pieces) - (r.x // 3) * (r.y // 3)) / sum(r.pieces) for r in regions_too_small]
print(
    "For these regions, the gap to fit in 3x3 blocks is between",
    " {:2.0f}% and {:2.0f}% too many pieces".format(min(gaps) * 100, max(gaps) * 100)
)

Found 6 pieces
Found 1000 regions
Of which 421 regions have too many pieces to fit in 3x3 blocks
For these regions, the gap to fit in 3x3 blocks is between  27% and 37% too many pieces


In [12]:
points_to_slots = []
for region in regions_too_small:
    numpoints = sum(region.pieces[i] * sum(pieces[i].grid) for i in range(len(region.pieces)))
    numslots = region.x * region.y
    points_to_slots.append(numpoints / numslots)

mini = min(points_to_slots)
maxi = max(points_to_slots)

print(
    f"Regions that are too small on the 3x3 basis must accomondate between {mini} and {maxi} points per slot"
)

Regions that are too small on the 3x3 basis must accomondate between 1.0004081632653061 and 1.0024489795918368 points per slot


## Analysis

- In all cases, one or two regions are enough.
- Only 421 regions out of a 1000 *possibly* need a second region.
- for these 421 regions, the ratio of points to slots is superior to 1. So, it's impossible to fit them in a single region.
- So, 1000 - 421 = 579 regions can fit all their presents.
